src/zmqports.cpp

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2014 The Bitcoin developers
+// Distributed under the MIT/X11 software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "zmqports.h"
+
+#if ENABLE_ZMQ
+#include <zmq.h>
+#endif
+
+#include "core/block.h"
+#include "core/transaction.h"
+#include "version.h"
+#include "streams.h"
+#include "util.h"
+#include "netbase.h"
+#include "ui_interface.h"
+#include "json/json_spirit_value.h"
+
+using namespace json_spirit;
+using namespace std;
+
+// Global state
+bool fZMQPub = false;
+
+#if ENABLE_ZMQ
+
+// ZMQ related file scope variables
+static void *zmqContext;
+static void *zmqPubSocket;
+
+// Internal utility functions
+static void ZMQPublishBlock(const CBlock &blk);
+static void ZMQPublishTransaction(const CTransaction &tx);
+
+static void zmqError(const char *str)
+{
+  LogPrint("ZMQ error: %s, errno=%s\n", str, zmq_strerror(errno));
+}
+
+// Called at startup to conditionally set up ZMQ socket(s)
+void ZMQInitialize(const std::string &endp)
+{
+  zmqContext = zmq_init(1);
+  if (!zmqContext) {
+    zmqError("Unable to initialize ZMQ context");
+    return;
+  }
+
+  zmqPubSocket = zmq_socket(zmqContext, ZMQ_PUB);
+  if (!zmqPubSocket) {
+    zmqError("Unable to open ZMQ pub socket");
+    return;
+  }
+
```

Comment:
I think `ZMQ_SNDHWM` should be set here to configurable value (at least through configure.ac) this provides the option to the user on when to start dropping messages. the default value of zero means *a queue as large as the available memory* which is dangerous because networks could be slow or the production of messages is faster than the network I/O output -- a precursor to a memory overflow crash.

`ZMQ_LINGER` is another important socket property to set here, the default value of -1 (infinite linger period on socket shutdown when there are queued messages to be sent) will prevent the application from shutting down gracefully. A configurable value (or a reasonable one) should be set to prevent this scenario. The linger period starts after a call to `zmq_close` on a socket or `zmq_ctx_term` on a context, once it expires all queued messages will be dropped.